Creating ACL 2 4 python3 test module

j-ororke opened the pull request:
Creating ACL 2 4 python3 test module:
- Adding python3 ACL_2_4 test module 
- This uses base branch of Amine's PR #[38263](https://github.com/project-chip/connectedhomeip/pull/38263) to create this test module: This is a follow-up PR to that PR from Amine, once Amine's PR is merged this one can come out of WIP state.

#### Testing
WSL - Windows Subsystem Linux environment with pyvenv

Alami-Amine committed:
Testing removing one of the template functions

Alami-Amine committed:
Modifying how ACL Attribute is written: making it use a non-empty initial ReplaceAll list

Alami-Amine committed:
WriteClient: Encoding as many items as possible into a single list as part of ReplaceAll item operation

Alami-Amine committed:
fixing failures

Alami-Amine committed:
Fix to Error Handling in User-Label Cluster: return RESOURCE_EXHAUSTED when the number of items in list exceeds kMaxUserLabelListLength

Alami-Amine committed:
chip-repl: support for preencoded attributes

Alami-Amine committed:
Fixing ACL 2.3 and 2.5

Alami-Amine committed:
Validating ACLs Upfront: This is done to make sure we reject the whole list if any of the entries within it are non conforming

Alami-Amine committed:
more yaml test changes

restyled-commits committed:
Restyled by prettier-yaml

Alami-Amine committed:
Fix Unit Testing Cluster (Test Cluster Server) to process ReplaceAll operations

Alami-Amine committed:
fix PutPreencodedAttribute

Alami-Amine committed:
Cleaning code

Alami-Amine committed:
More Code Clean Up

Alami-Amine committed:
Unit Test Cleanup

Alami-Amine committed:
more cleanup

Alami-Amine committed:
Adding testcases to TestAccessControlCluster

Alami-Amine committed:
Fixing comments

Alami-Amine committed:
Making IsValid method in AccessControl Static

Alami-Amine committed:
Adding comments

Alami-Amine committed:
Make IsValid a member function of AccessControl::Entry

Alami-Amine committed:
Adding WriteClient class member mIsWriteRequestChunked

Alami-Amine committed:
making TestWriteChunkig more Robust

Alami-Amine committed:
removing Test modifications on ACL Extensions

Alami-Amine committed:
adding reference to issue in TODO comment

Alami-Amine committed:
integrating comments

Alami-Amine committed:
integrating comments

Alami-Amine committed:
Revert "removing Test modifications on ACL Extensions"

This reverts commit 055af80d6c71e40ddeef081b6046263bc06ef141.

j-ororke committed:
Adding ACL_2_4 python3 test module:
- Adding python3 ACL_2_4 test module
- This uses base branch of Amine's PR #38263 to create this test module: This is a follow-up PR to that PR from Amine, once Amine's PR is merged this one can come out of WIP state.

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updating TC_ACL_2_4 python3 test module:
- Changing expect clause for test step 29

j-ororke committed:
Updating TC_ACL_2_4:
- Resolving linting errors

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1731 @@
+                     "Result is SUCCESS, value is stored as F1"),
+            TestStep(3, "TH1 reads DUT Endpoint 0 AccessControl cluster ACL attribute",
+                     "Result is SUCCESS, value is list of AccessControlEntryStruct containing 1 element with admin entry N1"),
+            TestStep(4, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with 3 elements (N1 admin, N2 view, N3 operate)", "Result is SUCCESS"),
```
I don't see any mention of N2 or N3 in test plan

Alami-Amine replied:
Should we add more details to the TestStep? how does this work?
Because for the second entry, we have the following NodeIDs: [111, 222, 333, 444] and for third entry we have : [555, 666, 777, 888]

j-ororke replied:
I think that is a great idea to make that test step more detailed, using your idea in the above commit, how about something like this:
TestStep(4, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with 3 elements:\n" +
                    "1. Admin entry (N1, privilege: Administer, authMode: CASE)\n" +
                    "2. View entry (subjects: [111, 222, 333, 444], privilege: View, authMode: Group, targets: [{Cluster: 11}, {Endpoint: 22}])\n" +
                    "3. Operate entry (subjects: [555, 666, 777, 888], privilege: Operate, authMode: Group, targets: [{Cluster: 55}, {Endpoint: 66}])",
                    "Result is SUCCESS"),

Alami-Amine replied:
looks good to me!

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1731 @@
+            TestStep(6, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with modified entries (N1 admin, N2 manage, N3 admin)", "Result is SUCCESS"),
+            TestStep(7, "TH1 reads DUT Endpoint 0 AccessControl cluster ACL attribute",
+                     "Result is SUCCESS, value is list of AccessControlEntryStruct containing 3 elements with updated privileges"),
+            TestStep(8, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with updated targets", "Result is SUCCESS"),
```
```suggestion
            TestStep(8, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with updated entries", "Result is SUCCESS"),
```
In my opinion, Since we are also updating Privilege and AuthMode

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1731 @@
+                     "Result is SUCCESS, value is list of AccessControlEntryStruct containing 3 elements with updated targets"),
+            TestStep(10, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with null subjects", "Result is SUCCESS"),
+            TestStep(11, "TH1 reads DUT Endpoint 0 AccessControl cluster ACL attribute",
+                     "Result is SUCCESS, value is list of AccessControlEntryStruct containing 3 elements with null subjects"),
```
```suggestion
                     "Result is SUCCESS, value is list of AccessControlEntryStruct containing 3 elements with null subjects in the last 2 elements"),
```

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1731 @@
+                     "Result is SUCCESS, value is list of AccessControlEntryStruct containing 2 elements"),
+            TestStep(16, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with ProxyView privilege", "Result is SUCCESS"),
+            TestStep(17, "TH1 reads DUT Endpoint 0 AccessControl cluster ACL attribute",
+                     "Result is SUCCESS, value is list of AccessControlEntryStruct containing 2 elements with ProxyView privilege"),
```
```suggestion
                     "Result is SUCCESS, value is list of AccessControlEntryStruct containing 2 elements with ProxyView privilege for the last element"),
```

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1731 @@
+            TestStep(19, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with max subjects", "Result is SUCCESS"),
+            TestStep(20, "TH1 reads DUT Endpoint 0 AccessControl cluster ACL attribute",
+                     "Result is SUCCESS, value contains MAXSUBJECTS random node IDs"),
+            TestStep(21, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with specific subjects", "Result is SUCCESS"),
```
```suggestion
            TestStep(21, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with 4 valid CATs as subjects in the last entry", "Result is SUCCESS"),
```

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1731 @@
+                     "Result is SUCCESS, value contains MAXSUBJECTS random node IDs"),
+            TestStep(21, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with specific subjects", "Result is SUCCESS"),
+            TestStep(22, "TH1 reads DUT Endpoint 0 AccessControl cluster ACL attribute",
+                     "Result is SUCCESS, value contains the specified subjects"),
```
```suggestion
                     "Result is SUCCESS, value is list of AccessControlEntryStruct containing 2 elements with last element having 4 valid CATs as subjects),
```

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1731 @@
+                     "Result is SUCCESS, value contains MAXTARGETS targets"),
+            TestStep(26, "TH1 reads DUT Endpoint 0 AccessControl cluster AccessControlEntriesPerFabric attribute",
+                     "Result is SUCCESS, value is stored as MAXENTRIES"),
+            TestStep(27, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with MAXENTRIES elements", "Result is SUCCESS"),
```
```suggestion
            TestStep(27, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with a list of MAXENTRIES elements", "Result is SUCCESS"),
```

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1731 @@
+                     "Result is SUCCESS, value is stored as MAXENTRIES"),
+            TestStep(27, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with MAXENTRIES elements", "Result is SUCCESS"),
+            TestStep(28, "TH1 reads DUT Endpoint 0 AccessControl cluster ACL attribute",
+                     "Result is SUCCESS, value contains MAXENTRIES elements"),
```
```suggestion
                     "Result is SUCCESS, value is list of AccessControlEntryStruct containing MAXENTRIES elements"),
```

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1731 @@
+            asserts.assert_true(isinstance(entry.targets, Nullable),
+                                f"Entry {i} targets should be Nullable")
+
+        # First reset ACL to only admin entry
```
Good idea! :+1:

j-ororke replied:
Thank you!

Alami-Amine replied:
Welcome, will reference this ticket so we remember to add this to the Test Plan [#5052](https://github.com/CHIP-Specifications/chip-test-plans/pull/5052)

Alami-Amine replied:
- @j-ororke I have added your step to the PR in :https://github.com/CHIP-Specifications/chip-test-plans/pull/5061
- I did that as a seperate step (I think it is the right way?)
- So i moved up the numbering of all other steps

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1731 @@
+                     "Result is SUCCESS for first element, CONSTRAINT_ERROR for second element"),
+            TestStep(30, "TH1 reads DUT Endpoint 0 AccessControl cluster ACL attribute",
+                     "Result is SUCCESS, value contains only the admin entry"),
+            TestStep(31, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute testing Group auth mode for admin",
```
```suggestion
            TestStep(31, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute testing invalid combination of Group auth mode with admin privilege",
```

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1731 @@
+            Clusters.AccessControl.Structs.AccessControlEntryStruct(
+                privilege=Clusters.AccessControl.Enums.AccessControlEntryPrivilegeEnum.kOperate,
+                authMode=Clusters.AccessControl.Enums.AccessControlEntryAuthModeEnum.kCase,
+                subjects=[0xFFFFFFFD00000000],  # Invalid group ID range
```
```suggestion
                subjects=[0xFFFFFFFD00000000],  # Invalid CAT 
```

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1731 @@
+                [(0, acl_attribute(value=group_id_range_acl))]
+            )
+            asserts.assert_equal(result[0].Status, Status.ConstraintError,
+                                 "Write ACL with invalid group ID range should fail with CONSTRAINT_ERROR")
```
```suggestion
                                 "Write ACL with an invalid CAT as subject should fail with CONSTRAINT_ERROR")
```

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1731 @@
+        except Exception as e:
+            self.print_step("Got expected error for invalid group ID range", str(e))
+
+        # Step 37: Test fabric-scoped node ID (should fail)
```
change all "fabric-scoped node ID" with invalid Group Node ID

j-ororke replied:
Sorry about that, thank you for catching those, I think I mistyped in the above names for the ID's. 
This should now be updated to the correct names in the latest code changes.

Alami-Amine replied:
no issues! Thanks

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1731 @@
+                     "Result is SUCCESS for first element, CONSTRAINT_ERROR for second element"),
+            TestStep(41, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute testing invalid device type",
+                     "Result is SUCCESS for first element, CONSTRAINT_ERROR for second element"),
+            TestStep(42, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute testing endpoint with device type",
```
```suggestion
            TestStep(42, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute testing an invalid target containing both endpoint and device type fields",
```

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1731 @@
+                     "Result is SUCCESS for first element, CONSTRAINT_ERROR for second element"),
+            TestStep(42, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute testing endpoint with device type",
+                     "Result is SUCCESS for first element, CONSTRAINT_ERROR for second element"),
+            TestStep(43, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute testing all target fields",
```
```suggestion
            TestStep(43, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute testing an invalid target containing endpoint, device type and cluster fields at once",",
```

j-ororke committed:
Apply commit suggestions from code review from Amine

Updating to include commit suggestions to test steps and comments from @Alami-Amine. These suggestions made the code cleaner and more readable, thank you!

Co-authored-by: Amine Alami <43780877+Alami-Amine@users.noreply.github.com>

j-ororke committed:
Updating TC_ACL_2_4 python3 test module to resolve commits from Amine, making the code cleaner and better:
- Updating to change type of ID from "fabric-scoped node ID" with correct type "invalid Group Node ID"
- Updating test step 4 description to be more detailed.

restyled-commits committed:
Restyled by autopep8

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1735 @@
+            asserts.assert_true(isinstance(entry.targets, Nullable),
+                                f"Entry {i} targets should be Nullable")
+
+        # Step 21: Write ACL with specific subjects
+        self.step(21)
+        # Define specific test subject IDs
+        CAT1 = 0x434154_1  # 'CAT1' in hex
+        CAT2 = 0x434154_2  # 'CAT2' in hex
+        CAT3 = 0x434154_3  # 'CAT3' in hex
+        CAT4 = 0x434154_4  # 'CAT4' in hex
```
```suggestion
        # Step 21: Write ACL with 4 valid CATs as subjects
        self.step(21)
        # Define various Valid CATs taken from Test Spec
        CAT1 = 0xFFFFFFFDABCD0001  # 'CAT1' in hex
        CAT2 = 0xFFFFFFFDFFFF0001  # 'CAT2' in hex
        CAT3 = 0xFFFFFFFD071C1074  # 'CAT3' in hex
        CAT4 = 0xFFFFFFFD071C1074  # 'CAT4' in hex
```
```

j-ororke committed:
Update src/python_testing/TC_ACL_2_4.py

Changing hex values for CAT1-CAT4 in test step 21 as suggested by Amine who gathered these from the test spec

Co-authored-by: Amine Alami <43780877+Alami-Amine@users.noreply.github.com>

j-ororke committed:
Updating TC_ACL_2_4 python3 test module:
- Replacing duplicate CAT4 value in test step 21 and replaced it with value from Amine that did not replicate CAT3's value

j-ororke committed:
Updating TC_ACL_2_4 test module:
- Updating to match test steps in test plan PR: https://github.com/CHIP-Specifications/chip-test-plans/pull/5052

j-ororke committed:
Pulling in files from upstream master to make sure that everything is in sync

j-ororke committed:
Updating TC_ACL_2_4 python3 test module:
- Removing try except blocks
- Replacing self.print_step() with logging.info()

restyled-commits committed:
Restyled by autopep8

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1628 @@
+            TestStep(28, "TH1 reads DUT Endpoint 0 AccessControl cluster ACL attribute",
+                     "Result is SUCCESS, value is list of AccessControlEntryStruct containing MAXENTRIES elements"),
+            TestStep(29, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute testing PASE auth mode",
+                     "Result is SUCCESS for first element, CONSTRAINT_ERROR for second element"),
```
```suggestion
                     "Result is CONSTRAINT_ERROR.  Second invalid element is rejected "),
```

Alami-Amine replied:
I have just changed it in test plan as well

Alami-Amine replied:
FYI, as of now, we have two possibilities for the first entry (in this and all below Steps):

1. First Entry accepted (which is the case in legacy implementations)
2. First Entry rejected (if the entire list is rejected; which is the case after my changes).

j-ororke replied:
TC_ACL_2_3 test module PR #38493 has a similar implementation to make sure that scenario 2 is experienced.
In this test I believe we can use the method we have implemented here as the multiple structs entry with one invalid struct and one valid struct should get rejected is covered in the TC_ACL_2_3 test module already. 
Wherein during test step 16 for TC_ACL_2_3 test module there is a constraint error as the whole list is rejected due to there being 2 structs with the 2nd struct being invalid while the 1st struct is valid, thus making both structs entry being rejected and we experience the expected constraint error.

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1628 @@
+            TestStep(30, "TH1 reads DUT Endpoint 0 AccessControl cluster ACL attribute",
+                     "Result is SUCCESS, value contains only the admin entry"),
+            TestStep(31, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute testing invalid combination of Group auth mode with admin privilege",
+                     "Result is SUCCESS for first element, CONSTRAINT_ERROR for second element"),
```
```suggestion
                     "Result is CONSTRAINT_ERROR.  Second invalid element is rejected "),
```

j-ororke replied:
Thank you, I have updated the test steps after step 29 and iterated through the test steps to be in the correct order.

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1628 @@
+        asserts.assert_equal(len(read_admin_only), 1,
+                             "ACL should contain only admin entry after reset")
+
+        # Now test PASE auth mode (should fail)
```
Can you please put this as step 30, (see previous comment)

j-ororke replied:
Thank you, this has been updated in the test script

Alami-Amine commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1628 @@
+            TestStep(27, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with a list of MAXENTRIES elements", "Result is SUCCESS"),
+            TestStep(28, "TH1 reads DUT Endpoint 0 AccessControl cluster ACL attribute",
+                     "Result is SUCCESS, value is list of AccessControlEntryStruct containing MAXENTRIES elements"),
+            TestStep(29, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute testing PASE auth mode",
```
I have added a step 29 in test plan which involves resetting the ACLs, can you please add it, and move all below steps by one  count. Thanks!
```suggestion
            TestStep(30, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute testing PASE auth mode",
```

j-ororke replied:
Thank you, I have updated to match the new test step information for test steps 29 and 30 to match with the test plan.

Alami-Amine replied:
Thanks, just a small addition
```suggestion
            TestStep(29, "TH1 writes DUT Endpoint 0 AccessControl cluster ACL attribute with acl_original",
```

j-ororke commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1628 @@
+        asserts.assert_equal(result[0].Status, Status.ConstraintError,
+                             "Write ACL with PASE auth mode should fail with CONSTRAINT_ERROR")
+
+        # Step 30: Read and verify ACL after failed PASE write
+        self.step(30)
```
Thank you, I have updated the test steps to be in the correct order.

j-ororke committed:
Updated TC_ACL_2_4 python3 test module:
- Added new test step method to step 29
- Updated test steps after test step 29 include the test step ordering
- Updated test 29 and 30 to include the correct verbiage for the definitions and expected results

j-ororke committed:
Updated TC_ACL_2_4 test module:
- Actually updated TC_ACL_2_4 python3 test module to correct expected results verbiage for test steps 32-44

restyled-commits committed:
Restyled by autopep8

Alami-Amine commented:
@j-ororke should we also remove the YAML script?

Alami-Amine reviewed (commented):
Should we also remove the YAML Script?

j-ororke committed:
Update src/python_testing/TC_ACL_2_4.py

Co-authored-by: Amine Alami <43780877+Alami-Amine@users.noreply.github.com>

j-ororke committed:
Removing Test_TC_ACL_2_4.yaml script

cecille commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1630 @@
+        # Verify ACL contents match what we wrote
+        asserts.assert_equal(len(read_acl), 3, "ACL should contain exactly 3 entries")
+        # Verify each entry matches what we wrote
+        for i, entry in enumerate(read_acl):
```
Would it not work to just do 
asserts.assert_in(new_acl[0], read_acl, "whatever msg")
asserts.assert_in(new_acl[1], read_acl, "whatever msg") 
etc.

j-ororke replied:
Sorry, I had not thought about using asserts.assert_in() for this scenario. 
However, it works really well it appears and it reduced the number of lines in the test module by a lot and is much much cleaner now! 👍

j-ororke committed:
Updating TC_ACL_2_4 python3 test module:
- Replaced enumerate blocks with asserts.assert_in() checks to verify that structs are read correctly throughout the test
- Removed some unneeded comments in the test module

restyled-commits committed:
Restyled by autopep8

cecille commented on src/python_testing/TC_ACL_2_4.py:
```diff
@@ -0,0 +1,1351 @@
+        )
+        asserts.assert_equal(result[0].Status, Status.ConstraintError,
+                             "Write ACL with Group auth mode for admin should fail with CONSTRAINT_ERROR")
+
```
Can you please add the check in there that the ACL list matches [acl_original] - check should happen on all the CONSTRAINT_ERROR values.

j-ororke replied:
No problem, I will add that check into those steps.

j-ororke committed:
Updating TC_ACL_2_4 python3 test module:
- Added in validation of admin entry only to end of test steps 32-44
- Removed duplicate validation of admin entry only in test step 31

j-ororke committed:
Merge branch 'master' into update_TC_ACL_2_4_test

Alami-Amine commented:
- Backward Compatibility: Passed Tests when run against `chip-all-clusters-app` from Matter 1.0
